src/app/shared/components/course-form/course-form.component.ts

```diff
@@ -7,111 +7,167 @@ import {
+  ) {}
+
+  ngOnInit() {
+    if (this.isUpdate) this.updateInit();
+    else this.createInit();
+  }
 
-  constructor(public fb: FormBuilder) {
+  createInit() {
+    this.buildForm();
+    this.subscription = this.courseStore.authors$.subscribe(authors$ => {
+      this.allAuthors = authors$;
+      this.nonCourseAuthors = this.allAuthors.filter(
+        authorAllElement =>
+          !this.authors.value.some((authorElement: Author) => authorElement.id === authorAllElement.id)
+      );
+    });
+  }
+  updateInit() {
+    this.actionText = 'UPDATE COURSE';
+    this.subscription = this.courseStore.authors$.subscribe(authors$ => {
+      this.allAuthors = authors$;
+      this.nonCourseAuthors = this.allAuthors.filter(
+        authorAllElement =>
+          !this.courseInfo?.authors.some((authorElement: Author) => authorElement.id === authorAllElement.id)
+      );
+    });
     this.buildForm();
   }
 
   buildForm(): void {
     this.courseForm = this.fb.group({
-      title: ["", [Validators.required, Validators.minLength(2)]],
-      description: ["", [Validators.required, Validators.minLength(2)]],
+      title: [this.courseInfo?.title || '', [Validators.required, Validators.minLength(2)]],
+      description: [this.courseInfo?.description || '', [Validators.required, Validators.minLength(2)]],
       newAuthor: this.fb.group({
-        author: [
-          "",
-          [
-            Validators.required,
-            Validators.minLength(2),
-            this.authorValidator(),
-          ],
-        ],
+        author: ['', [Validators.minLength(2), this.authorValidator()]],
       }),
-      duration: ["", [Validators.required, Validators.min(1)]],
-      authors: this.fb.array([]), // FormArray
+      duration: [this.courseInfo?.duration || '', [Validators.required, Validators.min(1)]],
+      authors: this.fb.array(this.courseInfo?.authors || []), // FormArray
     });
   }
 
   addAuthorToCourse(author: Author): void {
     this.authors.push(this.fb.control(author));
-    this.nonCourseAuthors = this.nonCourseAuthors.filter(
-      (nonCourseAuthor) => nonCourseAuthor.id !== author.id,
-    );
+    this.nonCourseAuthors = this.nonCourseAuthors.filter(nonCourseAuthor => nonCourseAuthor.id !== author.id);
   }
   removeAuthorFromCourse(author: Author, arrayIndex: number): void {
     this.nonCourseAuthors = [...this.nonCourseAuthors, author];
     this.authors.removeAt(arrayIndex);
   }
   // Use the names `title`, `description`, `author`, 'authors' (for authors list), `duration` for the form controls.
   createCourse(): void {
-    return undefined;
+    const course: Omit<Course, 'id'> = {
+      title: this.title?.value,
+      description: this.description?.value,
+      duration: this.duration?.value,
+      authors: this.authors?.value.map((author: Author) => author.id),
+    };
+    this.courseStore.createCourse(course);
+  }
+  updateCourse(): void {
+    const course: Omit<Course, 'id'> = {
```

Comment:
We can safely assume that a course being edited has an id. You are passing the course object and the isUpdate boolean property to this component. How can you simplify thing here a bit?

Reply:
If you think of to check ActivatedRoute param ( if not skip the expalantion : ) , actually I have done this way before, but the current looks better for me : it was decided in the form component if it is and update or a create based on the activated route params. However it looks more efficient for me that we inject router only where needed. ( in edit-course component )  and get the id according to param just in the forms where we need. This way the shared courseform component is a bit lighter(no injected router in it and unncessary logic for add course). Maybe its better to keep shared components as light as possible. Ok the drawback are the input or inputs ( we could have only one as well ) but because it has not much Input I felt this is more optimal.